File: app/src/main/java/com/mgaetan89/showsrage/fragment/LogsFragment.java
package com.mgaetan89.showsrage.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.adapter.LogsAdapter;
import com.mgaetan89.showsrage.model.LogLevel;
import com.mgaetan89.showsrage.model.Logs;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.util.ArrayList;
import java.util.List;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class LogsFragment extends Fragment implements Callback<Logs>, SwipeRefreshLayout.OnRefreshListener {
	@Nullable
	private LogsAdapter adapter = null;

	@Nullable
	private TextView emptyView = null;

	@NonNull
	private final List<String> logs = new ArrayList<>();

	@Nullable
	private RecyclerView recyclerView = null;

	@Nullable
	private SwipeRefreshLayout swipeRefreshLayout = null;

	public LogsFragment() {
		this.setHasOptionsMenu(true);
	}

	@Override
	public void failure(RetrofitError error) {
		if (this.swipeRefreshLayout != null) {
			this.swipeRefreshLayout.setRefreshing(false);
		}

		error.printStackTrace();
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		this.onRefresh();
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);

		inflater.inflate(R.menu.logs, menu);

		int menuId = getMenuIdForLogLevel(this.getPreferredLogsLevel());

		if (menuId > 0) {
			menu.findItem(menuId).setChecked(true);
		}
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_logs, container, false);

		if (view != null) {
			this.emptyView = (TextView) view.findViewById(android.R.id.empty);
			this.recyclerView = (RecyclerView) view.findViewById(android.R.id.list);
			this.swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipe_refresh);

			if (this.recyclerView != null) {
				this.adapter = new LogsAdapter(this.logs);

				this.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
					@Override
					public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
						super.onScrolled(recyclerView, dx, dy);

						if (LogsFragment.this.swipeRefreshLayout != null) {
							LogsFragment.this.swipeRefreshLayout.setEnabled(!recyclerView.canScrollVertically(-1));
						}
					}
				});
				this.recyclerView.setAdapter(this.adapter);
				this.recyclerView.setLayoutManager(new LinearLayoutManager(this.getActivity()));
			}

			if (this.swipeRefreshLayout != null) {
				this.swipeRefreshLayout.setColorSchemeResources(R.color.accent);
				this.swipeRefreshLayout.setOnRefreshListener(this);
			}
		}

		return view;
	}

	@Override
	public void onDestroy() {
		this.logs.clear();

		super.onDestroy();
	}

	@Override
	public void onDestroyView() {
		this.emptyView = null;
		this.recyclerView = null;
		this.swipeRefreshLayout = null;

		super.onDestroyView();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item.getGroupId() == R.id.menu_logs_level) {
			return this.handleLogsLevelSelection(item);
		}

		return super.onOptionsItemSelected(item);
	}

	@Override
	public void onRefresh() {
		if (this.swipeRefreshLayout != null) {
			this.swipeRefreshLayout.setRefreshing(true);
		}

		SickRageApi.getInstance().getServices().getLogs(this.getPreferredLogsLevel(), this);
	}

	@Override
	public void success(Logs logs, Response response) {
		if (this.swipeRefreshLayout != null) {
			this.swipeRefreshLayout.setRefreshing(false);
		}

		this.logs.clear();

		if (logs != null) {
			this.logs.addAll(logs.getData());
		}

		if (this.logs.isEmpty()) {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.VISIBLE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.GONE);
			}
		} else {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.GONE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.VISIBLE);
			}
		}

		if (this.adapter != null) {
			this.adapter.notifyDataSetChanged();
		}
	}

	@Nullable
	/* package */ static LogLevel getLogLevelForMenuId(@IdRes int menuId) {
		switch (menuId) {
			case R.id.menu_debug:
				return LogLevel.DEBUG;

			case R.id.menu_error:
				return LogLevel.ERROR;

			case R.id.menu_info:
				return LogLevel.INFO;

			case R.id.menu_warning:
				return LogLevel.WARNING;
		}

		return null;
	}

	@IdRes
	/* package */ static int getMenuIdForLogLevel(@Nullable LogLevel logLevel) {
		if (logLevel != null) {
			switch (logLevel) {
				case DEBUG:
					return R.id.menu_debug;

				case ERROR:
					return R.id.menu_error;

				case INFO:
					return R.id.menu_info;

				case WARNING:
					return R.id.menu_warning;
			}
		}

		return 0;
	}

	@NonNull
	private LogLevel getPreferredLogsLevel() {
		SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this.getActivity());
		String logsLevelString = preferences.getString(Constants.Preferences.Fields.LOGS_LEVEL, Constants.Preferences.Defaults.LOGS_LEVEL.name());

		try {
			return LogLevel.valueOf(logsLevelString);
		} catch (IllegalArgumentException ignored) {
			return Constants.Preferences.Defaults.LOGS_LEVEL;
		}
	}

	private boolean handleLogsLevelSelection(MenuItem item) {
		LogLevel logLevel = getLogLevelForMenuId(item.getItemId());

		if (logLevel != null) {
			// Check the selected menu item
			item.setChecked(true);

			// Save the selected logs level
			SharedPreferences.Editor editor = PreferenceManager.getDefaultSharedPreferences(this.getActivity()).edit();
			editor.putString(Constants.Preferences.Fields.LOGS_LEVEL, logLevel.name());
			editor.apply();

			// Refresh the list of logs
			this.onRefresh();

			return true;
		}

		return false;
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/LogsFragment.kt
package com.mgaetan89.showsrage.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.firebase.jobdispatcher.FirebaseJobDispatcher
import com.firebase.jobdispatcher.GooglePlayDriver
import com.firebase.jobdispatcher.Trigger
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.activity.MainActivity
import com.mgaetan89.showsrage.adapter.LogsAdapter
import com.mgaetan89.showsrage.extension.getLogLevel
import com.mgaetan89.showsrage.extension.getLogs
import com.mgaetan89.showsrage.extension.getLogsAutoUpdateInterval
import com.mgaetan89.showsrage.extension.getLogsGroup
import com.mgaetan89.showsrage.extension.getPreferences
import com.mgaetan89.showsrage.extension.saveLogLevel
import com.mgaetan89.showsrage.extension.saveLogs
import com.mgaetan89.showsrage.model.LogEntry
import com.mgaetan89.showsrage.model.LogLevel
import com.mgaetan89.showsrage.model.Logs
import com.mgaetan89.showsrage.network.SickRageApi
import com.mgaetan89.showsrage.service.LogsAutoUpdateService
import io.realm.Realm
import io.realm.RealmChangeListener
import io.realm.RealmResults
import kotlinx.android.synthetic.main.fragment_logs.empty
import kotlinx.android.synthetic.main.fragment_logs.list
import kotlinx.android.synthetic.main.fragment_logs.swipe_refresh
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response

class LogsFragment : Fragment(), Callback<Logs>, RealmChangeListener<RealmResults<LogEntry>>, SwipeRefreshLayout.OnRefreshListener {
	private var groups: Array<String>? = null
	private var jobDispatcher: FirebaseJobDispatcher? = null
	private var logLevel: LogLevel? = null
	private lateinit var logs: RealmResults<LogEntry>
	private lateinit var realm: Realm

	init {
		this.setHasOptionsMenu(true)
	}

	override fun failure(error: RetrofitError?) {
		this.swipe_refresh?.isRefreshing = false

		error?.printStackTrace()
	}

	override fun onActivityCreated(savedInstanceState: Bundle?) {
		super.onActivityCreated(savedInstanceState)

		val activity = this.activity

		if (activity is MainActivity) {
			activity.displayHomeAsUp(false)
			activity.setTitle(R.string.logs)
		}

		this.onRefresh()
	}

	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
		super.onActivityResult(requestCode, resultCode, data)

		if (requestCode == REQUEST_CODE_FILTER) {
			if (resultCode == Activity.RESULT_OK) {
				this.groups = data?.getStringArrayExtra(Constants.Bundle.LOGS_GROUPS)

				this.list?.adapter = null

				this.getLogs(this.getLogLevel())
			}
		}
	}

	override fun onChange(logs: RealmResults<LogEntry>) {
		if (this.list?.adapter == null) {
			this.list?.adapter = LogsAdapter(this.logs)
		}

		if (this.logs.isEmpty()) {
			this.empty?.visibility = View.VISIBLE
			this.list?.visibility = View.GONE
		} else {
			this.empty?.visibility = View.GONE
			this.list?.visibility = View.VISIBLE
		}

		this.list?.adapter?.notifyDataSetChanged()
	}

	override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
		inflater?.inflate(R.menu.logs, menu)

		if (!this.realm.isClosed) {
			menu?.findItem(R.id.menu_filter)?.isVisible = this.realm.getLogsGroup().isNotEmpty()
		}

		val menuId = getMenuIdForLogLevel(this.getLogLevel())

		if (menuId > 0) {
			menu?.findItem(menuId)?.isChecked = true
		}
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
			= inflater.inflate(R.layout.fragment_logs, container, false)

	override fun onOptionsItemSelected(item: MenuItem?): Boolean {
		if (item?.groupId == R.id.menu_logs_level) {
			return this.handleLogsLevelSelection(item)
		}

		if (item?.itemId == R.id.menu_filter) {
			this.handleLogsGroupFilter()

			return true
		}

		return super.onOptionsItemSelected(item)
	}

	override fun onRefresh() {
		this.swipe_refresh?.isRefreshing = true

		SickRageApi.instance.services?.getLogs(this.getLogLevel(), this)
	}

	override fun onStart() {
		super.onStart()

		this.realm = Realm.getDefaultInstance()
		this.logs = this.realm.getLogs(this.getLogLevel(), this.groups, this)
		this.list?.adapter = LogsAdapter(this.logs)
		this.scheduleAutoUpdate()
	}

	override fun onStop() {
		this.jobDispatcher?.cancel(AUTO_UPDATE_JOB_TAG)

		if (this.logs.isValid) {
			this.logs.removeAllChangeListeners()
		}

		this.realm.close()

		super.onStop()
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		this.list?.addOnScrollListener(object : RecyclerView.OnScrollListener() {
			override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
				super.onScrolled(recyclerView, dx, dy)

				swipe_refresh?.isEnabled = !(recyclerView?.canScrollVertically(-1) ?: false)
			}
		})
		this.list?.layoutManager = LinearLayoutManager(this.activity)

		this.swipe_refresh?.setColorSchemeResources(R.color.accent)
		this.swipe_refresh?.setOnRefreshListener(this)
	}

	override fun success(logs: Logs?, response: Response?) {
		this.swipe_refresh?.isRefreshing = false

		val logEntries = logs?.data?.map(::LogEntry) ?: emptyList()

		Realm.getDefaultInstance().let {
			it.saveLogs(this.getLogLevel(), logEntries)
			it.close()
		}

		this.activity?.invalidateOptionsMenu()
	}

	private fun getLogLevel(): LogLevel {
		this.logLevel?.let {
			return it
		}

		this.context?.let {
			return it.getPreferences().getLogLevel()
		}

		return Constants.Defaults.LOG_LEVEL
	}

	private fun getLogs(logLevel: LogLevel) {
		if (this.logs.isValid) {
			this.logs.removeAllChangeListeners()
		}

		this.logs = this.realm.getLogs(logLevel, this.groups, this)
	}

	private fun handleLogsGroupFilter() {
		LogsFilterFragment.newInstance(this.groups)
			.show(this.childFragmentManager, "logs_filter")
	}

	private fun handleLogsLevelSelection(item: MenuItem?): Boolean {
		this.logLevel = getLogLevelForMenuId(item?.itemId)

		return this.logLevel?.let {
			// Check the selected menu item
			item?.isChecked = true

			// Save the selected logs level
			this.context?.getPreferences()?.saveLogLevel(it)

			// Update the list of logs
			this.list?.adapter = null

			this.getLogs(it)

			// Refresh the list of logs
			this.onRefresh()

			true
		} ?: false
	}

	private fun scheduleAutoUpdate() {
		val autoUpdateInterval = this.context?.getPreferences().getLogsAutoUpdateInterval()

		if (autoUpdateInterval > 0) {
			this.jobDispatcher = FirebaseJobDispatcher(GooglePlayDriver(this.context))

			this.jobDispatcher?.let {
				val tolerance = autoUpdateInterval * TOLERANCE_RATIO
				val job = it.newJobBuilder()
						.setRecurring(true)
						.setService(LogsAutoUpdateService::class.java)
						.setTag(AUTO_UPDATE_JOB_TAG)
						.setTrigger(Trigger.executionWindow(autoUpdateInterval, autoUpdateInterval + tolerance.toInt()))
						.build()

				it.schedule(job)
			}
		}
	}

	companion object {
		const val REQUEST_CODE_FILTER = 1

		private const val AUTO_UPDATE_JOB_TAG = "logs-auto-update-tag"
		private const val TOLERANCE_RATIO = 0.1

		internal fun getLogLevelForMenuId(menuId: Int?) = when (menuId) {
			R.id.menu_debug -> LogLevel.DEBUG
			R.id.menu_error -> LogLevel.ERROR
			R.id.menu_info -> LogLevel.INFO
			R.id.menu_warning -> LogLevel.WARNING
			else -> null
		}

		internal fun getMenuIdForLogLevel(logLevel: LogLevel?) = when (logLevel) {
			LogLevel.DEBUG -> R.id.menu_debug
			LogLevel.ERROR -> R.id.menu_error
			LogLevel.INFO -> R.id.menu_info
			LogLevel.WARNING -> R.id.menu_warning
			else -> 0
		}

		fun newInstance() = LogsFragment()
	}
}
